src/optiga/optiga.c

```diff
@@ -963,6 +1259,164 @@ static int _verify_metadata(
+    if (res != OPTIGA_UTIL_SUCCESS) {
+        return res;
+    }
+    return res_clear;
+}
+}
+
+int optiga_init_new_password(const char* password)
+{
+    // Set new hmac key.
+    uint8_t new_hmac_key[32] = {0};
+    _ifs->random_32_bytes(new_hmac_key);
+    optiga_lib_status_t res = _optiga_util_write_data_sync(
+        _util, OID_HMAC, OPTIGA_UTIL_ERASE_AND_WRITE, 0x00, new_hmac_key, sizeof(new_hmac_key));
+    if (res != OPTIGA_UTIL_SUCCESS) {
+        util_log("failed updating the hmac key: %x", res);
+        return res;
+    }
+
+    // Set new symmetric key.
+    optiga_key_id_t keyid = OPTIGA_KEY_ID_SECRET_BASED;
+    res = _optiga_crypt_symmetric_generate_key_sync(
+        _crypt, OPTIGA_SYMMETRIC_AES_256, OPTIGA_KEY_USAGE_ENCRYPTION, false, &keyid);
+    if (res != OPTIGA_UTIL_SUCCESS) {
+        util_log("failed updating the sym key: %x", res);
+        return res;
+    }
+
+    uint8_t password_secret[32] = {0};
+    _ifs->random_32_bytes(password_secret);
+
+    res = _optiga_util_write_data_sync(
+        _util,
+        OID_PASSWORD_SECRET,
+        OPTIGA_UTIL_ERASE_AND_WRITE,
+        0x00,
+        password_secret,
+        sizeof(password_secret));
+    if (res != OPTIGA_UTIL_SUCCESS) {
+        return res;
+    }
+
+    uint8_t password_salted_hashed[32] = {0};
+    UTIL_CLEANUP_32(password_salted_hashed);
+    if (!salt_hash_data(
+            (const uint8_t*)password,
+            strlen(password),
+            "optiga_password",
+            password_salted_hashed)) {
+        return 1;
+    }
+
+    res = _set_password(
+        password_secret,
+        sizeof(password_secret),
+        password_salted_hashed,
+        sizeof(password_salted_hashed));
+    if (res) {
+        return res;
+    }
+    return 0;
+}
+
+bool optiga_reset_keys(void)
+{
+    // This resets the OID_AES_SYMKEY and OID_HMAC keys, as well as the OID_PASSWORD_SECRET and
+    // OID_PASSWORD keys. A password is needed because updating the OID_PASSWORD key requires
+    // auth using the OID_PASSWORD_SECRET key, but any password is fine for the purpose of resetting
+    // the keys.
+    return optiga_init_new_password("") == 0;
+}
+
+static int _optiga_verify_password(const char* password, uint8_t* password_secret_out)
+{
+    uint8_t password_salted_hashed[32] = {0};
+    UTIL_CLEANUP_32(password_salted_hashed);
+    if (!salt_hash_data(
+            (const uint8_t*)password,
+            strlen(password),
+            "optiga_password",
+            password_salted_hashed)) {
+        return 1;
+    }
+
+    optiga_lib_status_t res =
+        _authorize(OID_PASSWORD, password_salted_hashed, sizeof(password_salted_hashed));
+    if (res != OPTIGA_LIB_SUCCESS) {
+        goto cleanup;
+    }
+
+    uint16_t password_secret_size = 32;
+    res = _optiga_util_read_data_sync(
+        _util, OID_PASSWORD_SECRET, 0, password_secret_out, &password_secret_size);
+    if (res != OPTIGA_LIB_SUCCESS || password_secret_size != 32) {
```

Comment:
bug: the password size error does not set `res` to an error and so the function could succeed while it was supposed to fail.

Reply:
fixed